src/base/octets.rs

```diff
@@ -313,6 +313,86 @@ impl<'a, A: Array<Item = u8>> OctetsRef for &'a SmallVec<A> {
 }
 
 
+//------------ OctetsFrom ----------------------------------------------------
+
+/// Convert a type from one octets type to another.
+///
+/// This trait allows creating a value of a type that is generic over an
+/// octets sequence from an identical value using a different type of octets
+/// sequence.
+///
+/// This is different from just `From` in that the conversion may fail if the
+/// source sequence is longer than the space available for the target type.
+pub trait OctetsFrom<Source>: Sized {
+    /// Performs the conversion.
+    fn octets_from(source: Source) -> Result<Self, ShortBuf>;
+}
+
+
+impl<'a, Source: AsRef<[u8]> + 'a> OctetsFrom<&'a Source> for &'a [u8] {
+    fn octets_from(source: &'a Source) -> Result<Self, ShortBuf> {
+        Ok(source.as_ref())
```

Comment:
This blanket implementation is not enough. Bytes for example only supports conversion from `&'static [u8]`. So converting from `&Record` now requires cloning the record and converting each member using the owned value. Perhaps the blanket implementation could use OctetsBuilder which can convert any `Source: AsRef<[u8]>` into `Octets`?

Reply:
But isn’t that rather a missing `impl OctetsFrom<&'_ Record<...>> for Record<...>`? Or am I misunderstanding you?

Reply:
Ah yes you're right, I misread this, I thought it was implemented for any `AsRef<[u8]>`.

Reply:
It think when doing from a `&Record<O>` to a `Record<OO>`, we should go via `&O: OctetsRef`. So perhaps the trait bound should be `where &O: OctetsRef, OO: OctetsFrom<&O>` (plus lifetimes, of course). There might be a blanket impl in there somewhere, but it probably won’t work because of specialization rules.

Perhaps explore this in a different PR? Just so we get some progress here?